core/nightshadev2/src/nightshade.rs

```diff
@@ -0,0 +1,265 @@
+
+    /// Create new State with increased confidence using some proof
+    fn increase_confidence(&self, proof: SignedState) -> Self {
+        Self {
+            bare_state: BareState {
+                endorses: self.bare_state.endorses,
+                confidence0: self.bare_state.confidence0 + 1,
+                confidence1: self.bare_state.confidence1,
+            },
+            proof0: Some(proof),
+            proof1: self.proof1.clone(),
+        }
+    }
+
+    fn can_commit(&self) -> bool {
+        self.bare_state.confidence0 >= self.bare_state.confidence1 + COMMIT_THRESHOLD
+    }
+
+    fn verify(&self) -> bool {
+        true
+    }
+
+    fn get_signature(&self) -> BLSSignature {
+        0
+    }
+
+    fn endorses(&self) -> usize {
+        self.bare_state.endorses
+    }
+}
+
+impl PartialOrd for State {
+    fn partial_cmp(&self, other: &State) -> Option<Ordering> {
+        self.bare_state.partial_cmp(&other.bare_state)
+    }
+}
+
+impl Ord for State {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.bare_state.cmp(&other.bare_state)
+    }
+}
+
+
+fn update_confidence1(state0: &mut State, state1: &State){
+    // When this function is called it MUST hold that state0 >= state1. i.e.
+    // assert_eq!(state1 > state0, false);
+
+    if state0.endorses() != state1.endorses() {
+        if state1.bare_state.confidence0 > state0.bare_state.confidence1 {
+            state0.bare_state.confidence1 = state1.bare_state.confidence0;
+            state0.proof1 = state1.proof0.clone();
+        }
+    }
+}
+
+pub struct Nightshade {
+    owner_id: AuthorityId,
+    num_authorities: usize,
+    states: Vec<State>,
+    best_state_counter: usize,
+    seen_bare_states: HashSet<BareState>,
+    commited: bool,
+}
+
+impl Nightshade {
+    fn new(owner_id: AuthorityId, num_authorities: usize) -> Self {
+        let mut states = vec![];
+
+        for i in 0..num_authorities {
+            states.push(State::new(i));
+        }
+
+        Self {
+            owner_id,
+            num_authorities,
+            states,
+            best_state_counter: 1,
+            seen_bare_states: HashSet::new(),
+            commited: false,
+        }
+    }
+
+    fn update_state(&mut self, authority_id: AuthorityId, state: State) -> NSResult {
+        if commited {
+            return NSResult::Updated(None)
+        }
+
+        // Verify this BareState only if it has not been successfully verified previously
+        if !self.seen_bare_states.contains(&state.bare_state) {
+            if state.verify() {
+                self.seen_bare_states.insert(state.bare_state.clone());
+            } else {
+                return NSResult::Error("Not valid state".to_string());
+            }
+        }
+
+        if state.bare_state > self.states[authority_id].bare_state {
+            self.states[authority_id] = state.clone();
+
+            // We always take the best state seen so far
+            if state > self.states[self.owner_id] {
```

Comment:
same case as with update_confidence1:
if state[owner_id] is (B,4,4) and state[authority_id] is (A,5,3), new state[owner_id] should become (A,5,4)
(and publishing (B,4,4) and (A,5,3) by the same person should also be considered malicious behavior and potentially slashable)

Reply:
Fixed [here](https://github.com/nearprotocol/nearcore/pull/549/commits/ac553deead51f68bfae835e97283dfbd60ae9725#diff-eb6ef29898d59b2505e1e8a1c6dfab43R237)
There was also a bug handling `self.best_state_counter`.